request-manager/spec/parser.go

```diff
@@ -26,112 +24,95 @@ func ParseSpec(specFile string, logFunc func(string, ...interface{})) (Specs, er
 
 	return spec, nil
 }
 
 // Read all specs file in indicated specs directory.
-// `logFunc` is a Printf-like function used to log warning(s) should they occur.
+// 'logFunc' is a Printf-like function used to log warning(s) should they occur.
 // Errors are returned, not logged.
-func Parse(specsDir string, logFunc func(string, ...interface{})) (Specs, error) {
+func ParseSpecsDir(specsDir string, logFunc func(string, ...interface{})) (Specs, error) {
 	specs := Specs{
-		Sequences: map[string]*SequenceSpec{},
+		Sequences: map[string]*Sequence{},
 	}
 
+	failedFiles := []string{}
+	seqFile := map[string]string{} // sequence name --> file it was first seen in
 	err := filepath.Walk(specsDir, func(path string, info os.FileInfo, err error) error {
 		if err != nil {
 			return err
 		}
-		if info.IsDir() || !strings.HasSuffix(info.Name(), ".yaml") {
+		if info.IsDir() || !strings.HasSuffix(strings.ToLower(info.Name()), ".yaml") {
 			return nil
 		}
 		relPath, err := filepath.Rel(specsDir, path)
-		if err != nil {
-			logFunc("Warning: failed to get relative directory path for file %s: %s\n", path, err)
+		if err != nil { // if we can't get the relative path, just use the full path
+			relPath = path
 		}
 
 		spec, err := ParseSpec(path, logFunc) // logs warnings but not errors
 		if err != nil {
-			return fmt.Errorf("error reading spec file %s: %s", relPath, err)
+			failedFiles = append(failedFiles, fmt.Sprintf("%s: %s", relPath, err))
+			return nil
 		}
 
 		for name, spec := range spec.Sequences {
-			specs.Sequences[name] = spec
+			if _, ok := seqFile[name]; ok {
+				failedFiles = append(failedFiles, fmt.Sprintf("%s: sequence %s already seen in file %s", relPath, name, seqFile[name]))
+			} else {
+				specs.Sequences[name] = spec
+				seqFile[name] = relPath
+			}
 		}
 
 		return nil
 	})
 	if err != nil {
-		return specs, fmt.Errorf("error reading spec files: %s", err)
+		return specs, fmt.Errorf("error traversing specs directory: %s", err)
+	}
+	if len(failedFiles) > 0 {
+		multiple := ""
+		if len(failedFiles) > 1 {
+			multiple = "s"
+		}
+		return specs, fmt.Errorf("error in file%s:\n%s", multiple, strings.Join(failedFiles, "\n"))
 	}
 
 	return specs, nil
 }
 
-// Runs checks on allSpecs.
-// `logFunc` is a Printf-like function used to log warnings and errors should they occur.
-// If any error is logged, this function returns an error.
-func RunChecks(allSpecs Specs, logFunc func(string, ...interface{})) error {
-	sequenceErrors := makeSequenceErrorChecks()
-	nodeErrors := makeNodeErrorChecks(allSpecs)
-	nodeWarnings := makeNodeWarningChecks()
-
-	var ret error
-
-	for _, sequence := range allSpecs.Sequences {
-		for _, sequenceCheck := range sequenceErrors {
-			if err := sequenceCheck.CheckSequence(*sequence); err != nil {
-				logFunc("Error: %s\n", err)
-				ret = checkFailed
-			}
-		}
+// Specs require some processing after we've loaded them, but before we run the checker on them.
+func ProcessSpecs(specs Specs) error {
```

Comment:
I think it'd be better to take a `*Specs` and return nothing, since I don't really want to copy all the specs. I mean I could, and it probably won't take noticeably longer, and it only happens on startup, but like why.

It's not clear to me how strongly you prefer things to be purely functional when possible, so I'm not sure which change you think is better.

Reply:
I don't have a super strong opinion in this case — this was mostly motivated by our recent reading of Clean Code, which basically says to return `Specs` in this case for ease of reading — but I'm not 100% sold on that so I'm ok with either option. If you do take a `*Specs` and don't return anything, I would probably mention in the comment that the func is modifying the specs passed in.